src/mesh/mesh_adaptation/anisotropic_mesh_adaptation.cpp

```diff
@@ -0,0 +1,483 @@
+}
+
+template<int dim, int nstate, typename real, typename MeshType>
+void AnisotropicMeshAdaptation<dim, nstate, real, MeshType> :: reconstruct_p2_solution()
+{
+    assert(dg->get_min_fe_degree() == 2);
+    pcout<<"Reconstructing p2 solution."<<std::endl;
+    dg->assemble_residual(true);
+    VectorType delU = dg->solution;
+    solve_linear(dg->system_matrix, dg->right_hand_side, delU, dg->all_parameters->linear_solver_param);
+    delU *= -1.0;
+    delU.update_ghost_values();
+    dg->solution += delU;
+    dg->solution.update_ghost_values();
+}
+
+template<int dim, int nstate, typename real, typename MeshType>
+void AnisotropicMeshAdaptation<dim, nstate, real, MeshType> :: compute_feature_based_hessian()
+{
+    assert(dg->get_min_fe_degree() == dg->get_max_fe_degree());
+    assert(dg->get_min_fe_degree() == 2);
+    pcout<<"Computing feature based Hessian."<<std::endl;
+    // Based on Loseille, A. and Alauzet, F. "Continuous mesh framework part II.", 2011. 
+    // Compute Hessian of the solution for now (can be changed to Mach number or some other sensor when required).
+    const auto mapping = (*(dg->high_order_grid->mapping_fe_field));
+    dealii::hp::MappingCollection<dim> mapping_collection(mapping);
+    const dealii::UpdateFlags update_flags = dealii::update_jacobian_pushed_forward_grads | dealii::update_inverse_jacobians;
+    dealii::hp::FEValues<dim,dim>   fe_values_collection_volume (mapping_collection, dg->fe_collection, dg->volume_quadrature_collection, update_flags);
+    PHiLiP::FEValuesShapeHessian<dim> fe_values_shape_hessian;
+    
+    std::vector<dealii::types::global_dof_index> dof_indices(max_dofs_per_cell);
+
+    for(const auto &cell : dg->dof_handler.active_cell_iterators())
+    {
+        if(! cell->is_locally_owned()) {continue;}
+        
+        const unsigned int cell_index = cell->active_cell_index();
+        const unsigned int i_fele = cell->active_fe_index();
+        const unsigned int i_quad = i_fele;
+        const unsigned int i_mapp = 0;
+        fe_values_collection_volume.reinit(cell, i_quad, i_mapp, i_fele);
+        const dealii::FEValues<dim,dim> &fe_values_volume = fe_values_collection_volume.get_present_fe_values();
+        
+        const unsigned int n_dofs_cell = fe_values_volume.dofs_per_cell; 
+        dof_indices.resize(n_dofs_cell);
+        cell->get_dof_indices(dof_indices);
+    
+        cellwise_hessian[cell_index] = 0;
+        const unsigned int iquad = get_iquad_near_cellcenter(fe_values_volume.get_quadrature());
+        fe_values_shape_hessian.reinit(fe_values_volume, iquad);
+        const dealii::FESystem<dim,dim> &fe_ref = fe_values_volume.get_fe();
+        
+        for(unsigned int idof = 0; idof<n_dofs_cell; ++idof)
+        {
+            const unsigned int icomp = fe_values_volume.get_fe().system_to_component_index(idof).first;
+            // Computing hesssian of solution at state 0. Might need to change it later.
+            if(icomp == 0)
```

Comment:
In some articles, the hessian of the Mach is evaluated instead of one of the states. Perhaps a future version could consider any scalar.

Reply:
Yes, for Euler's equation the Mach number might be better. I initially thought about a function taking in any scalar field but haven't implemented it yet. It could definitely be implemented in a future PR.

Reply:
Could you add add this to the list of issues. It could be something a new/summer student takes on.

Reply:
Sure, I have added it to the list of issues.